### PR TITLE
Jakartaee CDI for liquibase

### PR DESCRIPTION
**Pull request by xazap**
## Impact

- [ ] Bug fix (non-breaking change which fixes expected existing functionality)
- [x] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description

For a Jakarta EE 10 project I wanted to use Liquibase but ran into issues with the CDI bootstrap to apply the Liquibase configuration. The existing liquibase-cdi submodule from release 4.18.0 cannot be used with JEE 9 or newer since the package names have been renamed from javax.* to jakarta.*

I decided to port the existing liquibase-cdi submodule to a liquibase-jakarta-cdi submodule for which this PR is the result. No new tests were added, but all existings unit tests are ported and run successfully. Note that the existing CDI module is not removed as it may still be used by other projects.

##  Comments on the ported module

- pom.xml:
    - Upgraded to CDI API 3.0.0.
    - Upgraded to Weld 4.0.3.Final, the reference implementation of CDI 3.0 mandated by JEE 9.
    - New artifact name "liquibase-jakarta-cdi" so that both legacy CDI and Jakarta CDI can co-exist.

- imports:
    - Refactored all CDI javax.* imports to their jakarta.* equivalents.

- package naming:
    - Used base package name of "liquibase.integration.jakarta.cdi" to differentiate from legacy CDI.
    
- CDIBootstrap.java:
    - Refactored CDILiquibase instance creation to use an InjectionTargetFactory to produce an InjectionTarget, as is now mandated by CDI 1.1+
    - Renamed bootstrap extension in META-INF/services from "javax.enterprise.inject.spi.Extension" to "jakarta.enterprise.inject.spi.Extension"

- CDILiquibase.java:
    - No code changed, added @Dependent to clarify the example and fixed some Javadoc rendering.
        
- beans.xml:
    - Updated schema reference to CDI version 3.0 as mandated by JEE 9.
    - Now using portable CDI grammar in favour of non-portable, Weld-specific grammar to exclude classes from scanning.
    
- CDILiquibaseTest.java:
   - Refactored deprecated instantiation of WeldContainer.
   - Auto-close WeldContainer after use.
    
- SchemesTreeBuilderTest.java:
   - Refactored testWithDependencies() and testMixed() since newer Weld distributions no longer provide com.google.common.base.* packages.
   - Minor refactoring to silence raw type compiler warnings.

## Things to be aware of

The Liquibase master branch sets the project version to 0-SNAPSHOT. This is unfortunate as the resulting filenames don't qualify for automatic module naming, see [ModuleFinder Javadoc](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/module/ModuleFinder.html#of(java.nio.file.Path...)). 

Since relying on filenames for module names is bad practice anyway (module name changes if the filename changes), I put "Automatic-Module-Name" entries in the manifests of liquibase-core and liquibase-jakarta-cdi to have proper stable module names. This maintains Java 8 compatibility (the manifest entry is ignored in Java 8) while also allowing newer runtimes to instrument the Java Platform Module System.

## Additional Context

Successfully tested resulting liquibase-core and liquibase-jakarta-cdi JARs in my project using a Payara Micro (JEE10) container to setup MariaDB.

### CODE REVIEW

**Comment by xazap** [comment]
I just noticed issue #1242. This PR should fix that.

**Comment by nvoxland** [comment]
**Initial/Pre-Review Thoughts**

The renaming of the javax -> jakarta packages is such a hassle.... 



Questions I have:
- Is there a way to share code between the two versions like we did with the jakarta vs javax servlet code?
- Will it correctly release the new module to maven central?
- Is `liquibase-jarkata-cdi` the right name to release it under on maven?  `liquibase-cdi-jakarta` alphabetically support nicer? A different term for what's different vs "regular" cdi than "jakarta"? Or good as it is?

Potential risks:
- Without a good testing environment, we have to trust it works. But, we know it doesn't work now so anything is an improvement :)

**Comment by xazap** [comment]
> The renaming of the javax -> jakarta packages is such a hassle....

Agreed!

>     * Is there a way to share code between the two versions like we did with the jakarta vs javax servlet code?

The ability to share code will be limited to classes that do not import either javax or jakarta CDI classes. I only found three candidates that could possibly be shared:
- CDILiquibaseConfig.java
- CyclicDependencyException
- DependencyNotFoundException

>     * Will it correctly release the new module to maven central?

I used the same Maven setup (parent/child) as the existing CDI module to integrate it into the build. If the new module inherits the repo credentials in the same way to release and deploy to maven central then it should work? Not tested.

>     * Is `liquibase-jarkata-cdi` the right name to release it under on maven?  `liquibase-cdi-jakarta` alphabetically support nicer? A different term for what's different vs "regular" cdi than "jakarta"? Or good as it is?

`liquibase-cdi-jakarta` is indeed a bit nicer. Alternatively it could be named after the CDI version like `liquibase-cdi-30`. Referencing version 3.0 of the CDI spec implies the use of the jakarta namespace.

> Potential risks:
> 
>     * Without a good testing environment, we have to trust it works. But, we know it doesn't work now so anything is an improvement :)

Well, the `CDILiquibaseTest` proves it works against Weld, which is the reference implementation for CDI 3.0. Any JEE9/JEE10 container correctly implementing CDI 3.0 should therefore work. I personally also tested against Payara full server and Payara Micro.

**Comment by filipelautert** [comment]
Hello @xazap ! Thanks again for the code, I'm working on reviewing it. Some comments:

1. I vote for `liquibase-cdi-jakarta` too, otherwise we will tie ourselves to speficid cdi version. Could you rename the module?
2. There is one sonar "bug"  that I guess was already present on the original jee code. Do you think we can handle it? -> https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&types=BUG&pullRequest=3642&id=liquibase&open=AYV-kOIat9dT63iyIk95
3. After that I'll run it using Wildfly 27 so we will have manually tested on 2 different servers.

**Comment by xazap** [comment]
>     1. I vote for `liquibase-cdi-jakarta` too, otherwise we will tie ourselves to speficid cdi version. Could you rename the module?
Done!

>     2. There is one sonar "bug"  that I guess was already present on the original jee code. Do you think we can handle it? -> https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&types=BUG&pullRequest=3642&id=liquibase&open=AYV-kOIat9dT63iyIk95

Looking at the code I  found that the InterruptedException is actually handled as advised. It turns out that this is a known false-positive in SonarQube, see https://sonarsource.atlassian.net/browse/SONARJAVA-4406

>     3. After that I'll run it using Wildfly 27 so we will have manually tested on 2 different servers.

Great!

**Comment by filipelautert** [comment]
Hello @xazap , thanks for the updates. 
I added some clean ups to the code and I think it's good to go (I applied the same clean up to liquibase-cdi too). 

To test it, I created a sample jakarta application and executed it on Wildfly 27.0.1 - it works fine! The code is located at https://github.com/filipelautert/liquibase-jakarta-ee-sample if anyone wants to try it. See important logs below:

```
23:13:36,357 WARN  [org.jboss.weld.Validator] (Weld Thread Pool -- 7) WELD-001473: jakarta.enterprise.inject.spi.Bean implementation liquibase.integration.jakarta.cdi.CDIBootstrap$1@61ec08c2 declared a normal scope but does not implementjakarta.enterprise.inject.spi.PassivationCapable. It won't be possible to inject this bean into a bean with a passivating scope (@SessionScoped, @ConversationScoped). This can be fixed by assigning the Bean implementation a unique id by implementing the PassivationCapable interface.
23:13:36,456 INFO  [liquibase.integration] (MSC service thread 1-5) Booting Liquibase [local build]
23:13:36,553 INFO  [liquibase.database] (MSC service thread 1-5) Set default schema name to PUBLIC
23:13:36,800 INFO  [liquibase.lockservice] (MSC service thread 1-5) Successfully acquired change log lock
23:13:37,146 INFO  [liquibase.changelog] (MSC service thread 1-5) Creating database history table with name: PUBLIC.DATABASECHANGELOG
23:13:37,151 INFO  [liquibase.changelog] (MSC service thread 1-5) Reading from PUBLIC.DATABASECHANGELOG
etc
```


Besides some code is duplicated to liquibase-cdi, I think that in a future release  we can drop liquibase-cdi and keep this one, so it's fine to keep it duplicated for now.

**Review by filipelautert — APPROVED** [review]
this pr introduces a new oss module to liquibase, `liquibase-cdi-jakarta` . It is an upgrade to `liquibase-cdi` and allows liquibase to be executed with recent versions of java application servers such as Wildfly 27 .
Liquibase-cdi still exists as the migration from javaee to jakarta ee will be slow, but someday in the future it can be dropped.

File: pom.xml [hunk 1]
```diff
@@ -272,7 +273,7 @@
                 <artifactId>maven-resources-plugin</artifactId>
                 <version>3.3.0</version>
                 <configuration>
-                    <encoding>UTF-8</encoding>
+                    <encoding>${project.build.sourceEncoding}</encoding>
```
Comment (filipelautert) [inline pom.xml:276]: This should not be changed, as it is the main pom.

[the remark this answers is not in the document]
Comment (filipelautert) [reply pom.xml]: Done!

File: liquibase-jakarta-cdi/src/test/java/liquibase/integration/jakarta/cdi/SchemesCDIConfigBuilderTest.java [hunk 2]
```diff
@@ -0,0 +1,290 @@
+package liquibase.integration.jakarta.cdi;
```
Comment (XDelphiGrl) [inline liquibase-jakarta-cdi/src/test/java/liquibase/integration/jakarta/cdi/SchemesCDIConfigBuilderTest.java]: Excellent tests - readable and comprehensive! Thanks!

File: liquibase-jakarta-cdi/src/test/java/liquibase/integration/jakarta/cdi/SchemesTreeBuilderTest.java [hunk 3]
```diff
@@ -0,0 +1,285 @@
+package liquibase.integration.jakarta.cdi;
```
Comment (XDelphiGrl) [inline liquibase-jakarta-cdi/src/test/java/liquibase/integration/jakarta/cdi/SchemesTreeBuilderTest.java]: Excellent tests - readable and comprehensive! Thank you!

File: liquibase-jakarta-cdi/src/main/resources/schema.template.xml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<databaseChangeLog
+        xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xmlns:ora="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-2.0.xsd
```
Comment (XDelphiGrl) [inline liquibase-jakarta-cdi/src/main/resources/schema.template.xml]: Please update the dbchangelog xsd version to `dbchangelog-latest.xsd` unless there is a reason 2.0 is required.

**Review by XDelphiGrl — APPROVED** [review]
This PR ports the Liquibase CDI integration from JavaX to Jakarta. Use this port of Liquibase CDI in application server environments running with JEE v9 and later. The Liquibase CDI Javax integration remains available and unchanged by this port.
* Test harness execution passing.
* Functional tests N/A (there are no CDI-specific functional tests).
* Existing tests ported from Liquibase CDI (JavaX) to Liquibase CDI Jakarta 
* No additional testing required.

APPROVED

**Review by suryaaki2 — APPROVED** [review]

**Review by filipelautert — APPROVED** [review]

**Pull request merged**